docs/orchestrator.8
.\"                                      Hey, EMACS: -*- nroff -*-
.TH ORCHESTRATOR 8 "July 29, 2020"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
orchestrator \- real-time container orchestrator
.SH SYNOPSIS
.B orchestrator
.RI "[ \-bBdDfFkmPqv ] [\-a " proc " ] [\-c " clock " ] [\-C " cgrp " ] \
[\-i " intv " ] [\-l " loop " ] [\-n " cmd " ] [\-p " prio " ] \
[\-\-policy=<policy>] [\-r " time " ] [\-\-rr " slice-time " ] [\-s " cmd "] \ 
[\-S " alg "] [\-w " time " ] [config.json]

.\" .SH DESCRIPTION
.\" This manual page documents briefly the
.\" .B orchestrator commands.
.\" .PP
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, respectively.
.\" \orchestrator\fP is a program that...
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes ('\-\-'). The specified config file defines container
orchestration and resource scheduling details. By default the software is set to static 
allocation, which means all task's resources are either preconfigured in the
[config.json] file and changed accordingly, or left as-is.
.br
A summary of options is included below.
.\" For a complete description, see the Info files.
.TP
.B \-a [PROC-SPLIT], \-\-affinity[=PROC-SPLIT]
Divide the CGroups of containers and system betweeo 0-[PROC-SPLIT-1] and 
[PROC-SPLIT]-MAX_CPU.
All containers will run on the latter control group, while systems processes will
be moved to the former, where possible
.br
The cpu numbers are the same as shown in the 
.I processor
field in /proc/cpuinfo.  See numa(3) for more information on specifying CPU sets.
.TP
.B \-A, \-\-adaptive
Enable adaptive scheduling algorithm. With this flag enabled, configured
task with affinity
.I -1
(all CPUs enabled, preference CPU 1) or smaller, get a precomputed affinity mask 
depending on free allocations and are then dynamically slotted on entry. The allocation
algorithm uses period affinity and utilization factor as selection criterion.
.TP
.B \-b, \-\-bind
bind non-real-time processes of the same container to the settings of the real-time
task settings of the container. For example, if a container hosts a non-real-time
process which launches a real-time thread, this flag will bind the non-real-time
main process to the same affinity and resource limit settings.
.TP
.B \-B, \-\-blind
enables the blind run option, such that the pre-run environment setup does not try to force any run-time environment change.
.TP
.B \-c CLOCK, \-\-clock=CLOCK
select clock source for timers and delays
.br
0 = CLOCK_MONOTONIC (default)
.br
1 = CLOCK_REALTIME
.br
2 = CLOCK_PROCESS_CPUTIME_ID
.br
3 = CLOCK_THREAD_CPUTIME_ID
.TP
.B -C [CGROUP]
Sets the PID detection mode to CGroup with sub-group named CGROUP. By default, the
CGROUP is set to 'docker/'. When changing, remind ending the group name with a slash.
.TP
.B \-d, \-\-dflag
set the deadline overshoot scheduling flag for qualifying SCHED_DEADLINE listed
tasks. TThis permits the task to reclaim possible unused real-time bandwidth beyond
the allotted amount.
.TP
.B \-D
sets the dry-run flag. No kernel or environment parameters will be changed but only actual values reported.
.TP
.B \-f
force execution with critical parameters, and enables changes that might influence other services on the system. This might be needed when, e.g. the period for the
update task in SCHED_DEADLINE is set to 1ms. Such a small period could cause a 
system overload and starve other processes.
.TP
.B \-F, \-\-ftrace
enables kernel function trace-based run-time statistics instead of the default
scheduler debug output interface. The latter is limited to a run-time refresh of 10ms.
.TP
.B \-i INTV, \-\-interval=INTV
Set the base interval for the update thread(s) for process data acquisition in
microseconds (default is 5000us).
.TP
.B \-k
with this flag, the data of PIDs that leave will be kept in memory, for statistical purposes only.
.TP
.B \-l LOOPS, \-\-loops=LOOPS
Set the number of intervals after which the kernel memory is cheked for new PIDs
mathing the detection pattern. Default loop numer is 100.
.TP
.B \-m, \-\-mlockall
Lock current and future memory allocations to prevent the orchestrator from being
paged out.
.TP
.B \-n [CMD]
use command line signaure detection. The optional CMD parameter specifies pid command
to look for when identifying cotainer daemons. Default command line parameter is 
'bash'.
.TP
.B \-p PRIO, \-\-prio=PRIO
Set the priority of the update thread. This value is real-time prio for SCHED_FIFO 
or SCHED_RR and  niceness for SCHED_OTHER, SCHED_NORMAL, SCHED_BATCH or SCHED_IDLE.
It has no meaning for SCHED_DEADLINE as there is no prioritized hierarichacal 
deadline scheduling available (yet).
.TP
.B \-P
enables thread scan for the matching PIDs to identify real-time children threads
as well.(used with -n)
.TP
.B \-\-policy=NAME
set the scheduler policy of the measurement threads
where NAME is one of: other, normal, batch, idle, fifo, rr, deadline
.TP
.B \-q
reduces verbosity where possible.
.TP
.B \-r TIME, \-\-runtime=TIME
sets the orchestrators runtime in seconds. Might come in handy when performing some
scripted test executions.
.TP
.B \-\-rr=TIME
The flag sets the slice value for Round-Robin scheduled tasks at startup.
.TP
.B \-s [CMD]
use shim PPID container detection. The optional CMD parameter specifies ppid command
to look for when identifying cotainer daemons. Default command line parameter is 
docker-containerd-shim.
.TP
.B \-S [NR], \-\-system[=NR]
selects the algorithm for the dynamic system scheduler (DSS). This scheduler tries to 
move containers when an overrun is imminent (Adaptive and Static do not do this). 
Montecarlo is not implemented yet.
.br
0 = simple system schedule based (default)
.br
1 = simple period based
.br
2 = Monte Carlo randomization
.TP
.B \-\-smi
enables interrupt counters to be read for each CPU. This to track the effect of 
uncontrollable interrupts;
.TP
.B \-v, \-\-verbose
adds verbosity if the binary has been compiled with the DEBUG flag
.TP
.B \-w TIME, \-\-wcet=TIME
sets the worst case execution time for the update thread, default is set to 100us.
(used with --policy=deadline)
.SH CONFIGURATION
The configuration file passed at CLI determines resource allocation and distrbution
of containers. The file is to be redacted in JSON format, strictly compliant.
By default,
.I config.json
will be taken as input file.
.TP
In the following and example of a configuration file

{                                      // open root object

    "global" : {
        "logdir" : "./",                          // logging directory - TODO
        "log_basename" : "orchestrator.txt",      // logging basename - TODO
        "prc_kernel" : "/proc/sys/kernel/",       // kernel settings vfs
        "sys_cgroup" : "/sys/fs/cgroup/",         // cgroup vfs for resource grouping
        "sys_cpu" :    "/sys/devices/system/cpu/",// cpu settings, ol vfs
        "cont_ppidc" : "containerd-shim",         // pid signature for container shim
        "cont_pidc" : "",                         // pid signature, opt, for filtering
        "cont_cgrp" : "docker/",                  // subdirectory for CGroups
        "priority" : 0,                           // priority of update thread
        "clock" : 0,                              // clock type to use
        "default_policy" : "SCHED_OTHER",         // scheduling policy for update thread
        "quiet" : 0,                              // quiet mode
        "setdflag" : 0,                           // sed dl-overrun flag  for deadline 
        "interval" : 5000,                        // update scan interval in us
        "dl_wcet": 100,                           // run-time budget for update thread
        "loops" : 100,                            // number of loops bf checking conts
        "runtime" : 0,                            // max run-time, 0 = infinite
        "psigscan" : 0,                           // parent signature scan 
        "trackpids" : 0,                          // keep left pids in stat
        "lock_pages" : 0,                         // lock pages for orchestrator
        "smi" : 0,                                // smi counter reading
        "rrtime" : 100,                           // slice time for rr scheduling, 100us
        "setaffinity" : "AFFINITY_UNSPECIFIED",   // affinity area for containers
        "affinity" : "1-2",                       // CPU affinity coma separated list
        "ftrace" : 0,                             // enable kernel function tracing 
        "ptresh" : 0.9,                           // dynamic scheduling probability thresh
    }
    "images" : [                    // list of images with matching containers & PIDs
    {                                  
        "imgid" : "4efcb6c965de",   // full or partial id (at least 12 characters) 
        "params" : { ... },         // real-time parameters - default to globals
        "res" : { ... },            // resouces for image, default to globals

        "cont" : [                  // list of containers based on this image
        {                                  
                "contid" : "4efcb6c965de",    // full or partial (at least 12 characters) of container id
                "params" : { ... }, // real-time parameters, default to image
                "res" : { ... },    // resouces to allocate, default to image
                "pids" : [{ 
                    "cmd" : "rt-app conf2.json",
                    "params" : { ... }, // real-time parameters, default to container
                    "res" : { ... },    // resouces to allocate, default to container
                 }],    // list of PID objects for this container
        }, ... ],

        "pids" : [                  // list of default PID objects for this image
        {
            "cmd" : "rt-app conf2.json",
            "params" : { ... },     // real-time parameters, default to image

src/orchestrator/orchestrator.c
// main settings and header file
#include "orchestrator.h"

// header files of launched threads
#include "prepare.h" // environment and configuration
#include "update.h"
#include "manage.h"
#include "adaptive.h"

// Default stuff, needed form main operation
#include <stdio.h>
#include <pthread.h>// used for thread management
#include <unistd.h> // used for POSIX XOPEN constants
#include <signal.h> // for SIGs, handling in main, raise in update

// Custom includes
#include "parse_config.h"	// header file of configuration parser
#include "kernutil.h"		// generic kernel utilities
#include "error.h"			// error and std error print functions
#include "cmnutil.h"		// common definitions and functions

// Things that should be needed only here
#include <sys/mman.h>		// memory lock
#include <getopt.h>			// command line parsing

/* --------------------------- Global variables for all the threads and programs ------------------ */

// containers has no lock-> verify single access only / RO
containers_t * contparm;	// container parameter settings
// prgset has no lock -> verify single access only / RO
prgset_t * prgset; 			// program settings structure

#ifdef DEBUG
// debug output file
	FILE  * dbg_out;
	FILE  * stats_out;
#endif

// mutex to avoid read while updater fills or empties existing threads
pthread_mutex_t dataMutex = PTHREAD_MUTEX_INITIALIZER;
// head of pidlist - PID runtime and configuration details
node_t * nhead = NULL;

// mutex to avoid read while updater fills or empties existing threads
pthread_mutex_t resMutex = PTHREAD_MUTEX_INITIALIZER; // UNUSED for now
// heads of resource allocations for CPU and Tasks
resTracer_t * rHead = NULL;

// -------------- LOCAL variables for all the functions  ------------------

// configuration read file
static char * config = "config.json";

// signal to keep status of triggers ext SIG
static volatile sig_atomic_t main_stop;

/* -------------------------------------------- DECLARATION END ---- CODE BEGIN -------------------- */


/// inthand(): interrupt handler for infinite while loop, help 
/// this function is called from outside, interrupt handling routine
/// Arguments: - signal number of interrupt calling
///
/// Return value: -
static void inthand (int sig, siginfo_t *siginfo, void *context){
	main_stop = 1;
}

/// display_dryrun(): Print dry-run mask settings for debug
///
/// Arguments: mask to print
///
/// Return value: -
static void display_dryrun(int msk){
	printf("INFO: Dry-run active with the following mask: ");
	if (MSK_DRYALL == msk)
		printf("SKIPALL");
	else
		for (int i = 1; i != 0; i<<=1){
			switch (msk & i){
			case MSK_DRYNOSMTOFF:
				printf("NOSMTOFF"); break;
			case MSK_DRYNOCPUGOV:
				printf("NOCPUGOV, ");// break;
	//		case MSK_DRYNOCPUQOS:
				printf("NOCPUQOS"); break;
			case MSK_DRYNORTTHRT:
				printf("NORTTHRT, "); //break;
	//		case MSK_DRYNORTSLCE:
				printf("NORTSLCE"); break;
			case MSK_DRYNOKTRDAF:
				printf("NOKTRDAF"); break;

			case MSK_DRYNOCGRPRT:
				printf("NOCGRPRT"); break;
			case MSK_DRYNOAFTY:
				printf("NOAFTY"); break;
			case MSK_DRYNOCPUPSH:
				printf("NOCPUPSH"); break;
			case MSK_DRYNOTSKPSH:
				printf("NOTSKPSH"); break;

			case MSK_DRYNOTRCNG:
				printf("NOTRCNG"); break;

			default:
				msk &= ~i;

			}
			if (msk & i){
				msk &= ~i;
				if (msk)
					printf(", ");
			}
		}
	printf("\n");
}

/// display_dryrun_help(): Print usage information
///
/// Arguments: exit with error?
///
/// Return value: -
static void display_dryrun_help(int error)
{
	(void)
	printf( "\nHere the values for '--dry-run=' mask.\n"
		    "The mask is a hex number where every bit disables a functionality interacting with the system.\n"
		    "Perform binary ands to compose the resulting hex mask.\n\n"

			"	0x%.4X		Do not disable SMT\n"
			"	0x%.4X		Do not change CPU governor\n"
			"	0x%.4X		Do not change CPU QoS setting -> same as above\n"
			"	0x%.4X		Do not change RT throttle\n"
			"	0x%.4X		Do not change RT RR slice\n"
			"	0x%.4X		Do not change kernel thread affinity\n"
			"\n"
			"	0x%.4X		Do not set Exclusive Control Groups ('root' for CGroup v2)\n"
			"	0x%.4X		Do not set affinity of container/task, CPU and memory\n"
			"	0x%.4X		Do not Hot-plug CPUs for IRQ push\n"
			"	0x%.4X		Do not push tasks into a new system group (only CGroup v1)\n"
			"\n"
			"	0x%.4X		Do not set tracing kernel debug -! expect malfunction\n"
			"\n"
			"	0x%.4X		Disable ALL\n"
			"\n",
			MSK_DRYNOSMTOFF, MSK_DRYNOCPUGOV, MSK_DRYNOCPUQOS, MSK_DRYNORTTHRT, MSK_DRYNORTSLCE, MSK_DRYNOKTRDAF,
			MSK_DRYNOCGRPRT, MSK_DRYNOAFTY, MSK_DRYNOCPUPSH, MSK_DRYNOTSKPSH, MSK_DRYNOTRCNG, MSK_DRYALL );

	if (error)
		exit(EXIT_FAILURE);
}

/// display_help(): Print usage information 
///
/// Arguments: exit with error?
///
/// Return value: - 
static void display_help(int error)
{
	(void)
	printf("Usage:\n"
	       "orchestrator <options> [config.json]\n\n"
	       "-a [NUM] --affinity        run container threads on specified cpu range,\n"
           "                           colon separated list\n"
	       "                           run system threads on remaining inverse mask list.\n"
		   "                           default: System=0, Containers=1-MAX_CPU\n"
	       "-A [NR] --adaptive[=NR]    activate Adaptive Static Schedule (ASS)\n"
	       "                           0 = Adaptive schedule \n"
		   "                           1 = Probabilistic adaptive schedule (default)\n"
		   "						       with period based reallocation\n"
#ifdef DEBUG
	       "-B       --blind           blind run (ignore environment preparation fails)\n"
	       "-c CLOCK --clock=CLOCK     select clock for measurement statistics\n"
	       "                           0 = CLOCK_MONOTONIC (default)\n"
	       "                           1 = CLOCK_REALTIME\n"
	       "                           2 = CLOCK_PROCESS_CPUTIME_ID\n"
	       "                           3 = CLOCK_THREAD_CPUTIME_ID\n"
#endif
           "-C [CGRP]                  use CGRP Docker directory to identify containers\n"
	       "                           optional CGRP parameter specifies base signature,\n"
           "                           default=%s\n"
	       "-d       --dflag           set deadline overrun flag for dl PIDs\n"
		   "-D                         dry run: suppress system changes/test only\n"
		   "         --dry-run=MASK    -\"-\"-  : set hex mask for dry-run mode\n"
	       "-f                         force execution with critical parameters\n"
	       "-F       --ftrace          start run-time analysis using kernel fTrace\n"
	       "-i INTV  --interval=INTV   base interval of update thread in us default=%d\n"
	       "-k                         keep track of ended PIDs\n"
	       "-l LOOPS --loops=LOOPS     number of loops for container check: default=%d\n"
	       "-m       --mlockall        lock current and future memory allocations\n"
	       "-n [CMD]                   use CMD signature on PID to identify containers\n"
	       "                           optional CMD parameter specifies base signature,\n"
           "                           default=%s\n"
	       "-p PRIO  --priority=PRIO   priority of the measurement thread:default=0\n"
	       "         --policy=NAME     policy of measurement thread, where NAME may be one\n"
	       "                           of: other, normal, batch, idle, deadline, fifo or rr.\n"
	       "-P                         with option -n, scan for children threads\n"
	       "-q       --quiet           print a summary only on exit\n"
	       "-r RTIME --runtime=RTIME   set a maximum runtime in seconds, default=0(infinite)\n"
	       "         --rr=RRTIME       set a SCHED_RR interval time in ms, default=100\n"
	       "-s [CMD]                   use shim PPID container detection.\n"
	       "                           optional CMD parameter specifies ppid command\n"
	       "-S [NR]  --system[=NR]     activate Dynamic System Schedule (DSS), alg NR \n"
	       "                           0 = Simple period based (default)\n"
	       "                           1 = Monte-Carlo bin (unsupported)\n"
#ifdef ARCH_HAS_SMI_COUNTER
           "         --smi             Enable SMI counting\n"
#endif
#ifdef DEBUG
	       "-v [LVL] --verbose[=LVL]   verbose output for debug purposes\n"
		   "                           level 1 = STDERR,\n"
		   "                           level 2 = Messages STDERR, stats file '/tmp/orch_stats.log'\n"
#endif
	       "-w       --wcet=TIME       WCET runtime for deadline policy in us, default=%d\n"
			, CGRP_DCKR, TSCAN, TDETM, CONT_PID, TWCET
		);
	if (error)
		exit(EXIT_FAILURE);

	// For --help query only

	display_dryrun_help(0);

	(void)
	printf("Report bugs to: f.hofer.phd@gmail.com\n"
	       "Project home page: <https://www.github.com/flhofer/real-time-containers/>\n");

	exit(EXIT_SUCCESS);
}

enum option_values {
	OPT_AFFINITY=1, OPT_ADAPTIVE, OPT_BLIND, OPT_CLOCK,
	OPT_DFLAG, OPT_DRYMASK, OPT_FTRACE, OPT_INTERVAL, OPT_LOOPS,
	OPT_MLOCKALL, OPT_NSECS, OPT_NUMA, OPT_PRIORITY, OPT_QUIET,
	OPT_RRTIME, OPT_RTIME, OPT_SYSTEM, OPT_SMI, OPT_VERBOSE,
	OPT_WCET, OPT_POLICY, OPT_HELP, OPT_VERSION
};

/// process_options(): Process commandline options 
///
/// Arguments: - structure with parameter set
///			   - passed command line variables
///			   - number of CPUs
///
/// Return value: -
static void process_options (prgset_t *set, int argc, char *argv[], int max_cpus)
{
	int error = 0;
	int option_index = 0;
	int optargs = 0;
#ifdef DEBUG
	dbg_out = fopen("/dev/null", "w");
	stats_out = fopen("/dev/null", "w");
	int verbose = 0;
#endif

	// preset configuration default values
	parse_config_set_default(set);

	for (;;) {
		/*
		 * Options for getopt
		 * Ordered alphabetically by single letter name
		 */
		static struct option long_options[] = {
			{"affinity",         optional_argument, NULL, OPT_AFFINITY},
			{"adaptive",         optional_argument, NULL, OPT_ADAPTIVE },
			{"blind",     		 no_argument,       NULL, OPT_BLIND },
			{"clock",            required_argument, NULL, OPT_CLOCK },
			{"dflag",            no_argument,		NULL, OPT_DFLAG },
			{"dry-run",          required_argument,	NULL, OPT_DRYMASK },
			{"ftrace",           no_argument,		NULL, OPT_FTRACE },
			{"interval",         required_argument, NULL, OPT_INTERVAL },
			{"loops",            required_argument, NULL, OPT_LOOPS },
			{"mlockall",         no_argument,       NULL, OPT_MLOCKALL },
			{"priority",         required_argument, NULL, OPT_PRIORITY },
			{"quiet",            no_argument,       NULL, OPT_QUIET },
			{"runtime",          required_argument, NULL, OPT_RTIME },
			{"rr",               required_argument, NULL, OPT_RRTIME },
			{"numa",             no_argument,       NULL, OPT_NUMA },
			{"smi",              no_argument,       NULL, OPT_SMI },
			{"system",           optional_argument, NULL, OPT_SYSTEM },
			{"version",			 no_argument,		NULL, OPT_VERSION},
			{"verbose",          optional_argument, NULL, OPT_VERBOSE },
			{"policy",           required_argument, NULL, OPT_POLICY },
			{"wcet",             required_argument, NULL, OPT_WCET },
			{"help",             no_argument,       NULL, OPT_HELP },
			{NULL, 0, NULL, 0}
		};
		int c = getopt_long(argc, argv, "a::A::bBc:C:dDfFhi:kl:mn::p:Pqr:s::S::v::w:",
				    long_options, &option_index);
		if (-1 == c)
			break;
		switch (c) {
		case 'a':
		case OPT_AFFINITY:
			if (NULL != optarg) {
				set->affinity = optarg;
				set->setaffinity = AFFINITY_USERSPECIFIED;
			} else if (optind<argc && atoi(argv[optind])) {
				set->affinity = argv[optind];
				optargs++;
				set->setaffinity = AFFINITY_USERSPECIFIED;
			// flag -a with no range = all
			} else {
				set->affinity = malloc(14);
				if (!set->affinity)
					err_exit("could not allocate memory!");
				(void)sprintf(set->affinity, "0-%d", max_cpus-1);
				set->setaffinity = AFFINITY_USEALL;
			}
			break;
		case 'A':
		case OPT_ADAPTIVE:
			set->sched_mode = SM_PADAPTIVE;
			if (NULL != optarg) {
				set->sched_mode= MIN(SM_ADAPTIVE + atoi(optarg), SM_PADAPTIVE);
			} else if (optind<argc) {
				set->sched_mode= MIN(SM_ADAPTIVE + atoi(argv[optind]), SM_PADAPTIVE);
				optargs++;
			}
			break;
#ifdef DEBUG
		case 'B':
		case OPT_BLIND:
			set->blindrun = 1; break;
		case 'c':
		case OPT_CLOCK:
			set->clocksel = atoi(optarg); break;
#endif
		case 'C':
			set->use_cgroup = DM_CGRP;
			if (NULL != optarg) {
				free(set->cont_cgrp);
				set->cont_cgrp = optarg;
			} else if (optind<argc) {
				free(set->cont_cgrp);
				set->cont_cgrp = argv[optind];
				optargs++;
			}

			/// -------------------- DOCKER & CGROUP CONFIGURATION
			// create Docker CGroup prefix
			parse_dockerfileprefix(set);
			break;
		case 'd':
		case OPT_DFLAG:
			set->setdflag = 1; break;
		case 'D':
			set->dryrun = 0xFFFF; break;
		case OPT_DRYMASK:
			set->dryrun = strtol(optarg, NULL, 16);
			if ((0 == set->dryrun) && errno){
				err_msg("Invalid dry-run mask '%s'!", optarg);
				display_dryrun_help(1);
			}
			break;
		case 'f':
			set->force = 1; break;
		case 'F':
		case OPT_FTRACE:
			set->ftrace = 1; break;
		case 'i':
		case OPT_INTERVAL:
			set->interval = atoi(optarg); break;
		case 'k':
			set->trackpids = 1; break;
		case 'l':
		case OPT_LOOPS:
			set->loops = atoi(optarg); break;
		case 'm':
		case OPT_MLOCKALL:
			set->lock_pages = 1; break;
		case 'n':
			set->use_cgroup = DM_CMDLINE;
			if (NULL != optarg) {
				free(set->cont_pidc);
				set->cont_pidc = optarg;
			} else if (optind<argc) {
				free(set->cont_pidc);
				set->cont_pidc = argv[optind];
				optargs++;
			}
			break;
		case 'p':
		case OPT_PRIORITY:
			set->priority = atoi(optarg);
			if (SCHED_FIFO != set->policy && SCHED_RR != set->policy) {
				warn(" policy and priority don't match: setting policy to SCHED_FIFO");
				set->policy = SCHED_FIFO;
			}
			break;
		case 'P':
			set->psigscan = 1; break;
		case 'q':
		case OPT_QUIET:
			set->quiet = 1; break;
		case 'r':
		case OPT_RTIME:
			if (NULL != optarg) {
				set->runtime = atoi(optarg);
			} else if (optind<argc && atoi(argv[optind])) {
				set->runtime = atoi(argv[optind]);
				optargs++;
			}
			break;
		case OPT_RRTIME:
			if (NULL != optarg) {
				set->rrtime = atoi(optarg);
			} else if (optind<argc && atoi(argv[optind])) {
				set->rrtime = atoi(argv[optind]);
				optargs++;
			}
			break;
		case 's':
			set->use_cgroup = DM_CNTPID;
			if (NULL != optarg) {
				free(set->cont_ppidc);
				set->cont_ppidc = optarg;
			} else if (optind<argc) {
				free(set->cont_ppidc);
				set->cont_ppidc = argv[optind];
				optargs++;
			}
			break;
		case 'S':
		case OPT_SYSTEM:
			// base Simple, limit to +1 = Monte Carlo;
			set->sched_mode = SM_DYNSIMPLE;
			if (NULL != optarg) {
				set->sched_mode += MIN(atoi(optarg), 1);
			} else if (optind<argc) {
				set->sched_mode += MIN(atoi(argv[optind]), 1);
				optargs++;
			}
			break;
#ifdef DEBUG
		case 'v':
		case OPT_VERBOSE: 
			verbose = 1; 
			if (NULL != optarg) {
				verbose = atoi(optarg);
			} else if (optind<argc) {
				verbose = atoi(argv[optind]);
				optargs++;
			}
			break;
#endif
		case OPT_VERSION:
			(void)printf("\nSource compilation date: %s\n", __DATE__);
			(void)printf("Copyright (C) 2019-24 Florian Hofer for\n"
						 " * Libera Università di Bolzano, Italy - Project I4S* - \"Industry for the Smart-*\"\n"
						 " * Siemens Corporate Technologies, Inc., Berkeley, CA, USA\n"
						 " * Università degli Studi di Bologna, Italy - Project PRIN 2022 - \"JOULE\"\n"
						 " * BI-REX - BIg data Research EXcellence competence center, Italy\n\n"

						 "License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>\n"
						 "This is free software: you are free to change and redistribute it.\n"
						 "There is NO WARRANTY, to the extent permitted by law.\n\n"

						 "Report bugs to: info@florianhofer.it\n"
						 "Project home page: <https://www.github.com/flhofer/real-time-containers/>\n");

			exit(EXIT_SUCCESS);
		case 'w':
		case OPT_WCET:
			set->update_wcet = atoi(optarg); break;
		case 'h':
		case OPT_HELP:
			display_help(0); break;
		case OPT_POLICY:
			if (optarg == NULL
				|| string_to_policy(optarg, &set->policy))
				err_exit("Invalid policy %s", optarg);

			break;
		case OPT_SMI:
#ifdef ARCH_HAS_SMI_COUNTER
			set->smi = 1;
#else
			err_exit("--smi is not available on your arch");
#endif
			break;
		}
	}

#ifdef DEBUG
	if (verbose) {
		fclose(dbg_out);
		dbg_out = stderr;

		fclose(stats_out);
		if (2 == verbose)
			stats_out = fopen("/tmp/orch_stats.log", "w");
		else
			stats_out = stderr;
	}
#endif

	// look for filename after options, we process only first
	if (optind+optargs < argc)
	{
	    config = argv[argc-1];
	}

	// always verify for configuration file -> segmentation fault??
	if ( access( config, F_OK )) {
		err_msg("configuration file '%s' not found", config);
		error = 1;
	}

	// create parameter structure
	if (!(contparm = calloc (1, sizeof(containers_t))))
		err_exit("Unable to allocate memory");

	if (!error)
		// parse json configuration
		parse_config_file(config, set, contparm);

	// fill missing default affinity values
	if (setDefaultAffinity(set))
		display_help(1);

	if (set->smi) { // verify this statements, I just put them all
		if (set->setaffinity == AFFINITY_UNSPECIFIED)
			err_exit("SMI counter relies on thread affinity");

		if (!has_smi_counter())
			err_exit("SMI counter is not supported "
			      "on this processor");
	}

	// check clock sel boundaries
	if (0 > set->clocksel || 3 < set->clocksel)
		error = 1;

	// check priority boundary
	if (0 > set->priority || 99 < set->priority)
		error = 1;

	// check detection mode and policy -> deadline does not allow fork!
	if (SCHED_DEADLINE == set->policy && (DM_CNTPID == set->use_cgroup || DM_CMDLINE == set->use_cgroup)) {
		warn("can not use SCHED_DEADLINE with PID detection modes: setting policy to SCHED_FIFO");
		set->policy = SCHED_FIFO;	
	}

	// deadline and high refresh might starve the system. require force
	if (SCHED_DEADLINE == set->policy && set->interval < 1000 && !set->force) {
		warn("Using SCHED_DEADLINE with such low intervals can starve a system. Use force (-f) to start anyway.");
		error = 1;
	}

	// check priority and policy match
	if (set->priority && (SCHED_FIFO != set->policy && SCHED_RR != set->policy)) {
		warn("policy and priority don't match: setting policy to SCHED_FIFO");
		set->policy = SCHED_FIFO;
	}

	// check policy with priority match 
	if ((SCHED_FIFO == set->policy || SCHED_RR == set->policy) && 0 == set->priority) {
		warn("defaulting real-time priority to %d", 10);
		set->priority = 10;
	}

	// if dryrun is set, display mask
	if (set->dryrun)
		display_dryrun(set->dryrun);

	// error present? print help message and exit
	if (error) {
		display_help(1);
	}
}

/// main(): main program.. setup threads and keep loop for user/system break
///
/// Arguments: - Argument values not defined yet
///
/// Return value: Exit code - 0 for no error - EXIT_SUCCESS
int main(int argc, char **argv)
{
	int max_cpus = sysconf(_SC_NPROCESSORS_ONLN);

	(void)printf("%s V %s\n", PRGNAME, VERSION);
	(void)printf("This software comes with no warranty. Please be careful\n");
	fflush(stdout);

	{ // pre-processing and configuration readout
		prgset_t *tmpset;
		if (!(tmpset = malloc (sizeof(prgset_t))))
			err_exit("Unable to allocate memory");

		process_options(tmpset, argc, argv, max_cpus);

		// gather actual information at startup, prepare environment
		if (prepareEnvironment(tmpset))
			display_help(1); // if it returns with error code, display help

		// NOW all is public!
		prgset = tmpset;
	}

	adaptPrepareSchedule(); // prepares masks, tracers and alike, useful for all active modes
	if (SM_ADAPTIVE <= prgset->sched_mode
			&& SM_DYNSIMPLE >= prgset->sched_mode){
		// adaptive scheduling active? Clean prepare, execute, useful for all active modes
		adaptPlanSchedule();
		adaptExecute();
	}
	if (SM_ADAPTIVE != prgset->sched_mode
			&& SM_PADAPTIVE != prgset->sched_mode)
		// free allocation info right away if we are not adaptive, dynamic allocation later
		adaptFree();

	pthread_t thrManage, thrUpdate;
	int32_t t_stat1 = 0; // we control thread status 32bit to be sure read is atomic on 32 bit -> sm on treads
	int32_t t_stat2 = 0; 
	int  iret1, iret2;

	/* Create independent threads each of which will execute function */ 
	if ((iret1 = pthread_create( &thrManage, NULL, thread_manage, (void*) &t_stat1))) {
		err_msg_n (iret1, "could not start update thread");
		t_stat1 = -1;
	}

	if ((iret2 = pthread_create( &thrUpdate, NULL, thread_update, (void*) &t_stat2))) {
		err_msg_n (iret2, "could not start management thread");
		t_stat2 = -1;
	}
#ifdef DEBUG
	(void)pthread_setname_np(thrManage, "manage");
	(void)pthread_setname_np(thrUpdate, "update");
#endif

	{ // setup interrupt handler block
		struct sigaction act;

		/* Use the sa_sigaction field because the handles has two additional parameters */
		/* The SA_SIGINFO flag tells sigaction() to use the sa_sigaction field, not sa_handler. */
		act.sa_handler = NULL; // On some architectures ---
		act.sa_sigaction = &inthand; // these are a union, do not assign both, -> first set null, then value
		act.sa_flags = SA_SIGINFO;

		/* blocking signal set during handler */
		sigemptyset(&act.sa_mask);
		sigaddset(&act.sa_mask, SIGINT);
		sigaddset(&act.sa_mask, SIGTERM);
		sigaddset(&act.sa_mask, SIGQUIT);
		sigaddset(&act.sa_mask, SIGHUP);

		act.sa_restorer = NULL;

		if ((sigaction(SIGINT, &act, NULL) < 0)		 // CTRL+C
			|| (sigaction(SIGTERM, &act, NULL) < 0)) // KILL termination or end of test
		{
			perror ("Setup of sigaction failed");
			exit(EXIT_FAILURE); // exit the software, not working
		}
	} // END interrupt handler block
	{ // signal blocking set

		 sigset_t set;
	   /* Block SIGQUIT and SIGHUP; other threads created by main()
		  will inherit a copy of the signal mask. */

	   // allow all to be handled by Main, except
	   if ( ((sigemptyset(&set)))
			   || ((sigaddset(&set, SIGQUIT)))	// used in manage thread
			   || ((sigaddset(&set, SIGHUP)))	// used in docker_link
			   || (0 != pthread_sigmask(SIG_BLOCK, &set, NULL))){
		   // INT signal, stop from main prg
			perror ("Setup of sigmask failed");
			exit(EXIT_FAILURE); // exit the software, not working
		}
	}

	// infinite loop until stop
	while (!main_stop && (t_stat1 > -1 || t_stat2 > -1)) {
		sleep (1);
	}

	// signal shutdown to threads if not set
	if (-1 < t_stat1) // only if not already done internally
		t_stat1 = -1;
	if (-1 < t_stat2) // only if not already done internally 
		t_stat2 = -1;

	// wait until threads have stopped
	if (!iret1) // thread started successfully
		iret1 = pthread_join( thrManage, NULL);
	if (!iret2)	// thread started successfully
		iret2 = pthread_join( thrUpdate, NULL); 

    // close and cleanup
    info("exiting safely");
    cleanupEnvironment(prgset);

#ifdef DEBUG
	if (dbg_out != stderr) {
		fflush(dbg_out);
		fclose(dbg_out);
	}
	if (stats_out != stderr) {
		fflush(stats_out);
		fclose(stats_out);
	}
#endif

    return EXIT_SUCCESS;
}
